diff --git a/Makefile b/Makefile
--- a/Makefile
+++ b/Makefile
@@ -160,7 +160,7 @@ ccflags-y += -DCONFIG_RTW_ANDROID=$(CONF
 endif
 
 ########################## Debug ###########################
-CONFIG_RTW_DEBUG = y
+CONFIG_RTW_DEBUG = n
 # default log level is _DRV_INFO_ = 4,
 # please refer to "How_to_set_driver_debug_log_level.doc" to set the available level.
 CONFIG_RTW_LOG_LEVEL = 0
@@ -196,7 +196,8 @@ CONFIG_LAYER2_ROAMING = y
 #bit0: ROAM_ON_EXPIRED, #bit1: ROAM_ON_RESUME, #bit2: ROAM_ACTIVE
 CONFIG_ROAMING_FLAG = 0x3
 ###################### Platform Related #######################
-CONFIG_PLATFORM_AUTODETECT = y
+CONFIG_PLATFORM_AUTODETECT = n
+CONFIG_PLATFORM_OPENWRT = y
 CONFIG_PLATFORM_I386_PC = n
 CONFIG_PLATFORM_ANDROID_X86 = n
 CONFIG_PLATFORM_ANDROID_INTEL_X86 = n
@@ -1413,6 +1414,34 @@ STAGINGMODDIR := /lib/modules/$(KVER)/ke
 endif
 # *** End AUTODETECT ***
 
+ifeq ($(CONFIG_PLATFORM_OPENWRT), y)
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/kernel.mk
+
+ccflags-y += -DCONFIG_LITTLE_ENDIAN \
+               -DCONFIG_IOCTL_CFG80211 \
+               -DRTW_USE_CFG80211_STA_EVENT \
+               -DRTW_IPS_MODE=0 \
+               -DRTW_LPS_MODE=0 \
+               -DRTW_WOW_LPS_MODE=0 \
+               -DCONFIG_TDLS \
+               -DCONFIG_RTW_MESH \
+               -DCONFIG_RTW_MULTI_AP \
+               -DRTW_P2P_GROUP_INTERFACE=1 \
+               -DCONFIG_FORCE_SW_CHANNEL_PLAN \
+               -DCONFIG_CONCURRENT_MODE \
+               -DCONFIG_IFACE_NUMBER=4 \
+               -DCONFIG_QOS_OPTIMIZATION \
+               -DRTW_REDUCE_SCAN_SWITCH_CH_TIME \
+               -Wno-error
+
+USER_MODULE_NAME := rtl$(MODULE_NAME)
+ARCH="$(LINUX_KARCH)"
+CROSS_COMPILE="$(TARGET_CROSS)"
+KSRC="$(LINUX_DIR)"
+KVER="$(LINUX_VERSION)"
+endif
+
 ifeq ($(CONFIG_PLATFORM_I386_PC), y)
 ccflags-y += -DCONFIG_LITTLE_ENDIAN
 ccflags-y += -DCONFIG_IOCTL_CFG80211 -DRTW_USE_CFG80211_STA_EVENT
diff --git a/os_dep/linux/ioctl_cfg80211.c b/os_dep/linux/ioctl_cfg80211.c
--- a/os_dep/linux/ioctl_cfg80211.c
+++ b/os_dep/linux/ioctl_cfg80211.c
@@ -465,7 +465,7 @@ u8 rtw_cfg80211_ch_switch_notify(_adapte
 		 */
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(RHEL88))
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0) && (LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0))) || (defined(RHEL89) && !defined(RHEL95))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0) && (LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0))) || (defined(RHEL89) && !defined(RHEL95)) && !defined(BUILD_OPENWRT)
 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, 0, false, 0);
 #else
 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, 0, false);
@@ -484,7 +484,7 @@ u8 rtw_cfg80211_ch_switch_notify(_adapte
 		goto exit;
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2) || defined(RHEL88))
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0) && (LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0))) || (defined(RHEL89) && !defined(RHEL95))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0) && (LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0))) || (defined(RHEL89) && !defined(RHEL95)) && !defined(BUILD_OPENWRT)
 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0, 0);
 #else
 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0);
@@ -3601,7 +3601,7 @@ static void cfg80211_rtw_abort_scan(stru
 }
 #endif
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0)) || defined(BUILD_OPENWRT)
 static int cfg80211_rtw_set_wiphy_params(struct wiphy *wiphy, int link_id, u32 changed)
 #else
 static int cfg80211_rtw_set_wiphy_params(struct wiphy *wiphy, u32 changed)
@@ -4590,7 +4590,7 @@ static int cfg80211_rtw_set_txpower(stru
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0))
 	struct wireless_dev *wdev,
 #endif
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0)) || defined(BUILD_OPENWRT)
 	int link_id,
 #endif
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 36)) || defined(COMPAT_KERNEL_RELEASE)
@@ -4655,10 +4655,10 @@ static int cfg80211_rtw_get_txpower(stru
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0))
 	struct wireless_dev *wdev,
 #endif
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0)) || defined(BUILD_OPENWRT)
 	int link_id,
 	unsigned int link_mask,
-#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0))
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)) || defined(BUILD_OPENWRT)
 	unsigned int link_id,
 #endif
 	int *dbm)
@@ -5422,7 +5422,7 @@ exit:
 }
 
 static int cfg80211_rtw_change_beacon(struct wiphy *wiphy, struct net_device *ndev,
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 7, 0)) || defined(RHEL94)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 7, 0)) || defined(RHEL94) || defined(BUILD_OPENWRT)
 	struct cfg80211_ap_update *info)
 #else
 	struct cfg80211_beacon_data *info)
@@ -5433,7 +5433,7 @@ static int cfg80211_rtw_change_beacon(st
 
 	RTW_INFO(FUNC_NDEV_FMT"\n", FUNC_NDEV_ARG(ndev));
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 7, 0)) || defined(RHEL94)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 7, 0)) || defined(RHEL94) || defined(BUILD_OPENWRT)
 	ret = rtw_add_beacon(adapter, info->beacon.head, info->beacon.head_len, info->beacon.tail, info->beacon.tail_len);
 #else
 	ret = rtw_add_beacon(adapter, info->head, info->head_len, info->tail, info->tail_len);
@@ -6924,7 +6924,7 @@ static void rtw_get_chbwoff_from_cfg8021
 #endif /* (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0)) */
 
 static int cfg80211_rtw_set_monitor_channel(struct wiphy *wiphy
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 13, 0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 13, 0)) || defined(BUILD_OPENWRT)
 	, struct net_device *dev
 	, struct cfg80211_chan_def *chandef
 #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0))
diff --git a/core/mesh/rtw_mesh.c b/core/mesh/rtw_mesh.c
--- a/core/mesh/rtw_mesh.c
+++ b/core/mesh/rtw_mesh.c
@@ -3021,11 +3021,15 @@ static void rtw_mpath_tx_queue_flush(_ad
 }
 
 #ifdef PLATFORM_LINUX /* 3.10 ~ 4.13 checked */
+#if defined(BUILD_OPENWRT)
+#include <linux/slab.h>
+#else
 #if defined(CONFIG_SLUB)
 #include <linux/slub_def.h>
 #elif defined(CONFIG_SLAB)
 #include <linux/slab_def.h>
 #endif
+#endif
 typedef struct kmem_cache rtw_mcache;
 #endif
 
diff --git a/os_dep/linux/os_intfs.c b/os_dep/linux/os_intfs.c
--- a/os_dep/linux/os_intfs.c
+++ b/os_dep/linux/os_intfs.c
@@ -2169,6 +2169,8 @@ int rtw_os_ndev_register(_adapter *adapt
 #ifdef CONFIG_RTW_NAPI
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(RHEL88))
 	netif_napi_add_weight(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT);
+#elif defined(BUILD_OPENWRT)
+	netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll);
 #else
 	netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT);
 #endif
